src/Agoda.Analyzers/AgodaCustom/AG0004AsyncAnalyzer.cs

```diff
@@ -1,93 +1,67 @@
-//                    if (invokeMethod.OriginalDefinition.Name.Equals("Wait"))
-//                    {
-//                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, location));
-//                        return;
-//                    }
-
-//                    // Checks if the WaitAny method is called within an async method then creates the diagnostic.
-//                    if (invokeMethod.OriginalDefinition.Name.Equals("WaitAny"))
-//                    {
-//                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, location));
-//                        return;
-//                    }
-
-//                    // Checks if the WaitAll method is called within an async method then creates the diagnostic.
-//                    if (invokeMethod.OriginalDefinition.Name.Equals("WaitAll"))
-//                    {
-//                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, location));
-//                        return;
-//                    }
-
-//                    // Checks if the Sleep method is called within an async method then creates the diagnostic.
-//                    if (invokeMethod.OriginalDefinition.Name.Equals("Sleep"))
-//                    {
-//                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, location));
-//                        return;
-//                    }
-
-//                    // Checks if the GetResult method is called within an async method then creates the diagnostic.     
-//                    if (invokeMethod.OriginalDefinition.Name.Equals("GetResult"))
-//                    {
-//                        context.ReportDiagnostic(Diagnostic.Create(Descriptor, location));
-//                        return;
-//                    }
-//                }
-
-//                var property = context.SemanticModel.GetSymbolInfo(context.Node).Symbol as IPropertySymbol;
-
-//                // Checks if the Result property is called within an async method then creates the diagnostic.
-//                if (property != null && property.OriginalDefinition.Name.Equals("Result"))
-//                {
-//                    context.ReportDiagnostic(Diagnostic.Create(Descriptor, location));
-//                    return;
-//                }
-//            }
-//        }
-//    }
-//}
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Agoda.Analyzers.Helpers;
+using System.Collections.Generic;
+
+namespace Agoda.Analyzers.AgodaCustom
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class AG0004AsyncAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "AG0004";
+
+        private readonly List<string> blockingMethodCallsOnTask = new List<string> { "Wait", "WaitAny", "WaitAll" };
```

Comment:
Is there no better way of doing this? My reading of this rule is that we are just checking the name of the class and method (eg. any class called `Task` with any method called `Wait`), not the actual type (eg. `typeof(Task)`).